### PR TITLE
[CLI] Add support for `ReduceLROnPlateau`

### LINKED ISSUE #10850: Support for ReduceLROnPlateau in CLI
## 🚀 Feature

From slack: it appears that since `ReduceLROnPlateau` doesn't subclass from `_LRScheduler` that it cannot be used with the amazing new optimizer/lr_scheduler loading in the YAML.

Try to use the new CLI with the following snippet in the YAML file:
```yaml
lr_scheduler:
  class_path: torch.optim.lr_scheduler.ReduceLROnPlateau
  init_args:
    factor: 0.1
    mode: min
    patience: 5
```
and it returned 

```
baseline_example.py: error: Configuration check failed :: Parser key "lr_scheduler": Value "{'class_path': 'torch.optim.lr_scheduler.ReduceLROnPlateau', 'init_args': {'factor': 0.1, 'mode': 'min', 'patience': 5}}" does not validate against any of the types in typing.Union[torch.optim.lr_scheduler.ChainedScheduler, torch.optim.lr_scheduler.ConstantLR, torch.optim.lr_scheduler.CosineAnnealingLR, torch.optim.lr_scheduler.CosineAnnealingWarmRestarts, torch.optim.lr_scheduler.CyclicLR, torch.optim.lr_scheduler.ExponentialLR, torch.optim.lr_scheduler.LambdaLR, torch.optim.lr_scheduler.LinearLR, torch.optim.lr_scheduler.MultiStepLR, torch.optim.lr_scheduler.MultiplicativeLR, torch.optim.lr_scheduler.OneCycleLR, torch.optim.lr_scheduler.SequentialLR, torch.optim.lr_scheduler.StepLR]:
  - "torch.optim.lr_scheduler.ReduceLROnPlateau" is not a subclass of ChainedScheduler
  - "torch.optim.lr_scheduler.ReduceLROnPlateau" is not a subclass of ConstantLR
  - "torch.optim.lr_scheduler.ReduceLROnPlateau" is not a subclass of CosineAnnealingLR
  - "torch.optim.lr_scheduler.ReduceLROnPlateau" is not a subclass of CosineAnnealingWarmRestarts
  - "torch.optim.lr_scheduler.ReduceLROnPlateau" is not a subclass of CyclicLR
  - "torch.optim.lr_scheduler.ReduceLROnPlateau" is not a subclass of ExponentialLR
  - "torch.optim.lr_scheduler.ReduceLROnPlateau" is not a subclass of LambdaLR
  - "torch.optim.lr_scheduler.ReduceLROnPlateau" is not a subclass of LinearLR
  - "torch.optim.lr_scheduler.ReduceLROnPlateau" is not a subclass of MultiStepLR
  - "torch.optim.lr_scheduler.ReduceLROnPlateau" is not a subclass of MultiplicativeLR
  - "torch.optim.lr_scheduler.ReduceLROnPlateau" is not a subclass of OneCycleLR
  - "torch.optim.lr_scheduler.ReduceLROnPlateau" is not a subclass of SequentialLR
  - "torch.optim.lr_scheduler.ReduceLROnPlateau" is not a subclass of StepLR
```

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

- Adds support for `--lr_scheduler=ReduceLROnPlateau`
- Allows customizing `configure_optimizers` in the CLI.

Fixes #10850 

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @carmocca @mauvilsa

### CODE REVIEW

File: pytorch_lightning/utilities/cli.py [hunk 1]
```diff
@@ -84,6 +85,15 @@ def __str__(self) -> str:
 LR_SCHEDULER_REGISTRY = _Registry()
 LR_SCHEDULER_REGISTRY.register_classes(torch.optim.lr_scheduler, torch.optim.lr_scheduler._LRScheduler)
 
+
+class ReduceLROnPlateau(torch.optim.lr_scheduler.ReduceLROnPlateau):
```
Comment (tchaton) [inline pytorch_lightning/utilities/cli.py:89]: Neat hack !

File: pytorch_lightning/utilities/cli.py [hunk 2]
```diff
@@ -690,12 +700,31 @@ def _parser(self, subcommand: Optional[str]) -> LightningArgumentParser:
         action_subcommand = self.parser._subcommands_action
         return action_subcommand._name_parser_map[subcommand]
 
-    def _add_configure_optimizers_method_to_model(self, subcommand: Optional[str]) -> None:
-        """Adds to the model an automatically generated ``configure_optimizers`` method.
+    @staticmethod
+    def configure_optimizers(
```
Comment (tchaton) [inline pytorch_lightning/utilities/cli.py:704]: Quick question. Do you believe we could support multiple optimizers, schedulers?

I was thinking about the following idea.
Assuming the following command line:

```py
--optimizer_1=Adam
--optimizer_2=.SGD
--lr_scheduler=Adam
```

and the following configure_optimizers method

```
def configure_optimizers(self):
     opt_gen = ...
     opt_dis = ...
     scheduler = ...(opt_gen)
     return [opt_gen, opt_dis], scheduler
```

We could call the real `configure_optimizers`, get the 2 optimizers and scheduler, and use those to make the same parameters & scheduler - optimizer association based on the command line.

Just an idea.

**Review by tchaton — APPROVED** [review]
LGTM !

File: docs/source/common/lightning_cli.rst [hunk 3]
```diff
@@ -892,45 +870,64 @@ Where the arguments can be passed directly through command line without specifyi
...
+
+.. code-block:: bash
+
+    $ python trainer.py fit \
+        --gen_optimizer=Adam \
+        --gen_optimizer.lr=0.01 \
+        --gen_discriminator=AdamW \
+        --gen_discriminator.lr=0.0001
+
+You can also use pass the class path directly, for example, if the optimizer hasn't been registered to the
```
Comment (awaelchli) [inline docs/source/common/lightning_cli.rst:920]: use pass

**Review by awaelchli — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: docs/source/common/lightning_cli.rst (hunk 3 above)
Comment (carmocca) [reply docs/source/common/lightning_cli.rst:920]:
```suggestion
You can also pass the class path directly, for example, if the optimizer hasn't been registered to the
```

File: pytorch_lightning/utilities/cli.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/utilities/cli.py:704]: One of the features of the CLI is removing the need for implementing `configure_optimizers` on the model for basic uses.

What you suggest goes against that, as the model implementation controls the CLI behavior.

This PR already makes the CLI behavior overridable. It doesn't support multiple optimizers at the moment but it could be supported.
